crawlerr/services/rag_service.py

```
@@ -73,6 +128,228 @@ def chunk_text(self, text: str, metadata: Dict[str, Any] = None) -> List[Dict[st
+            if end_token_idx < total_tokens:
+                next_start = max(0, end_token_idx - self.chunk_overlap_tokens)
+                if next_start <= previous_start:
+                    # No progress made, force advancement
+                    next_start = end_token_idx
+                start_token_idx = next_start
+            else:
+                break  # All tokens processed
+        
+        return chunks
+    
+    def _find_paragraph_boundary(self, search_text: str, ideal_end: int) -> Optional[int]:
```

Ruff (0.12.2)

Use `X | None` for type annotations

Convert to `X | None`

(UP045)